src/Effect/Exception.ss

```diff
@@ -0,0 +1,84 @@
+;; -*- mode: scheme -*-
+
+(library (Effect.Exception foreign)
+  (export showErrorImpl
+          error
+          errorWithCause
+          message
+          name
+          stackImpl
+          throwException
+          catchException)
+  (import (only (rnrs base) define lambda cond else quote)
+          (only (rnrs conditions) define-condition-type &condition condition? message-condition? condition-message)
+          (only (rnrs io ports) call-with-string-output-port)
+          (only (rnrs exceptions) with-exception-handler raise-continuable)
+          (only (chezscheme) format call/cc display-condition))
+
+  ; ----- Internal condition specific to Effect.Exception
+  ; Effect.Exception can catch other exceptions as well but
+  ; makes a distinction between exceptions thrown via its interface 
+  ; and all other exceptions so as to implement `errorWithCause` (parent exceptions).
+  (define-condition-type
+    &purescm-effect-exception
+    &condition
+    make-purescm-effect-exception
+    purescm-effect-exception? 
+    (message purescm-effect-exception-message)
+    (cause purescm-effect-exception-cause))
+
+  ; ----- Exported FFI implementation
+
+  (define showErrorImpl
+    (lambda (err)
+      (cond
+        ((condition? err)
+          (call-with-string-output-port
+            (lambda (p) (display-condition err p))))
+        (else
+          (format "Exception: ~s" err)))))
+
+  (define error
+    (lambda (msg)
+      (make-purescm-effect-exception msg '())))
+
+  (define errorWithCause
+    (lambda (msg)
+      (lambda (cause)
+        (make-purescm-effect-exception msg cause))))
```

Comment:
I think we should make this API work in the same way for any `Error`s (conditions) that might be thrown and not add special handling for errors thrown specifically using this API. Most conditions will not be created using this `errorWithCause` anyway. That's why I think it would be better if we didn't introduce our own condition type to carry the `cause`.

It seems to me that instead we could use a [compound condition](https://scheme.com/tspl4/exceptions.html) and just display them all in the `showErrorImpl`. Here we could create a compound condition containing the `cause` with

```scheme
(condition (make-message-condition msg) cause)
```

Then it seems we could get the entire stack of conditions using `simple-conditions` to be able to show all the conditions, including the "cause":

```
> (for-each (lambda (e) (display-condition e) (newline) (newline)) (simple-conditions (condition (make-message-condition "FAILED") cause)))
Exception: FAILED

Exception occurred with condition components:
  0. &warning

Exception: I'm the original cause

Exception occurred with condition components:
  0. &irritants: ()
```

Reply:
Good points, I like it more this way. I'll apply the changes now.